hazelcast/src/main/java/com/hazelcast/cache/impl/AbstractCacheRecordStore.java

```diff
@@ -317,15 +321,21 @@ public R processExpiredEntry(Data key, R record, long expiryTime, long now, Stri
         if (isStatisticsEnabled()) {
             statistics.increaseCacheExpiries(1);
         }
-        doRemoveRecord(key, source);
+        R removedRecord = doRemoveRecord(key, source);
+        Data keyEventData = toEventData(key);
+        Data recordEventData = toEventData(removedRecord);
+        onProcessExpiredEntry(key, removedRecord, expiryTime, now, source, origin);
         if (isEventsEnabled()) {
-            publishEvent(createCacheExpiredEvent(toEventData(key), toEventData(record),
+            publishEvent(createCacheExpiredEvent(keyEventData, recordEventData,
                                                  CacheRecord.EXPIRATION_TIME_NOT_AVAILABLE,
                                                  origin, IGNORE_COMPLETION));
         }
         return null;
     }
 
+    protected void onProcessExpiredEntry(Data key, R record, long expiryTime, long now, String source, String origin) {
```

Comment:
it would be better make it `abstract` method instead leaving it empty. The same applies to `onCreateRecordWithExpiryError` as well.